internal/middleware/memoryprotection/memoryprotection.go

```diff
@@ -0,0 +1,271 @@
+		SampleIntervalSeconds: 1,
+	}
+}
+
+// MemoryLimitProvider gets and sets the limit of memory usage.
+// In production, use DefaultMemoryLimitProvider.
+// For testing, use HardCodedMemoryLimitProvider.
+type MemoryLimitProvider interface {
+	Get() int64
+	Set(int64)
+}
+
+var (
+	_ MemoryLimitProvider = (*DefaultMemoryLimitProvider)(nil)
+	_ MemoryLimitProvider = (*HardCodedMemoryLimitProvider)(nil)
+)
+
+type DefaultMemoryLimitProvider struct{}
+
+func (p *DefaultMemoryLimitProvider) Get() int64 {
+	// SetMemoryLimit returns the previously set memory limit.
+	// A negative input does not adjust the limit, and allows for retrieval of the currently set memory limit
+	return debug.SetMemoryLimit(-1)
+}
+
+func (p *DefaultMemoryLimitProvider) Set(limit int64) {
+	debug.SetMemoryLimit(limit)
+}
+
+type HardCodedMemoryLimitProvider struct {
+	Hardcodedlimit int64
+}
+
+func (p *HardCodedMemoryLimitProvider) Get() int64 {
+	return p.Hardcodedlimit
+}
+
+func (p *HardCodedMemoryLimitProvider) Set(limit int64) {
+	p.Hardcodedlimit = limit
+}
+
+type MemoryAdmissionMiddleware struct {
+	config         Config
+	memoryLimit    int64 // -1 means no limit
+	metricsSamples []metrics.Sample
+	ctx            context.Context // to stop the background process
+
+	lastMemorySampleInBytes   *atomic.Uint64 // atomic because it's written inside a goroutine but can be read from anywhere
+	timestampLastMemorySample *atomic.Pointer[time.Time]
+}
+
+// New creates a new memory admission middleware with the given context.
+// Whe the context is cancelled, this middleware stops its background processing.
+func New(ctx context.Context, config Config, limitProvider MemoryLimitProvider, name string) MemoryAdmissionMiddleware {
+	am := MemoryAdmissionMiddleware{
+		config:                    config,
+		lastMemorySampleInBytes:   &atomic.Uint64{},
+		timestampLastMemorySample: &atomic.Pointer[time.Time]{},
+		memoryLimit:               -1, // disabled initially
+		ctx:                       ctx,
+	}
+
+	// Get the current GOMEMLIMIT
+	memoryLimit := limitProvider.Get()
+	if memoryLimit < 0 {
+		log.Info().Str("name", name).Msg("GOMEMLIMIT not set, memory protection disabled")
+		return am
+	}
+
+	if config.ThresholdPercent <= 0 {
+		log.Info().Str("name", name).Msg("threshold is non-positive; memory protection disabled")
+		return am
+	}
+
+	if config.SampleIntervalSeconds <= 0 {
+		log.Info().Str("name", name).Msgf("memory protection sample interval cannot be zero or negative; using default value of %q seconds", DefaultSampleIntervalSeconds)
+		am.config.SampleIntervalSeconds = DefaultSampleIntervalSeconds
+	}
+
+	am.memoryLimit = memoryLimit
+	am.metricsSamples = []metrics.Sample{
+		{Name: "/memory/classes/heap/objects:bytes"},
+	}
+
+	// Initialize with current memory usage
+	am.sampleMemory()
+
+	// Start background sampling with context
+	am.startBackgroundSampling()
+
+	log.Info().
+		Str("name", name).
+		Int64("memory_limit_bytes", memoryLimit).
```

Comment:
Maybe add another field with a humanized memory limit? makes it easier to parse, specially when you start using large machines